DownFall/Models/Board.swift

```diff
@@ -361,41 +340,39 @@ extension Board {
         
         //save the item
         guard case let TileType.item(item) = selectedTile.type,
-            var updatedTiles = transformation.endTiles else { return Transformation.zero }
-        
-        if let pp = playerPosition,
-            case let .player(data) = updatedTiles[pp].type {
-            
-            var items = data.carry.items
-            items.append(item)
-            let newCarryModel = CarryModel(items: items)
-            let playerData = EntityModel(originalHp: data.originalHp,
-                                         hp: data.hp,
-                                         name: data.name,
-                                         // Orend
-                // we have to reset attack here because the player has moved but the turn may not be over
-                // Eg: it is possible that there could be two or more monsters
-                // under the player and the player should be able to attack
-                attack: data.attack.resetAttack(),
-                type: data.type,
-                carry: newCarryModel,
-                animations: data.animations,
-                abilities: data.abilities)
+            var updatedTiles = transformation.endTiles,
+            let pp = playerPosition,
+            case let .player(data) = updatedTiles[pp].type
+            else { return Transformation.zero }
             
-            updatedTiles[pp.x][pp.y] = Tile(type: .player(playerData))
-        }
+        let newCarryModel = data.carry.earn(item.amount, inCurrency: item.type.currencyType)
+        let playerData = EntityModel(originalHp: data.originalHp,
+                                     hp: data.hp,
+                                     name: data.name,
+                                     // Orend
+            // we have to reset attack here because the player has moved but the turn may not be over
+            // Eg: it is possible that there could be two or more monsters
+            // under the player and the player should be able to attack
+            attack: data.attack.resetAttack(),
+            type: data.type,
+            carry: newCarryModel,
+            animations: data.animations,
+            abilities: data.abilities)
+        
+        updatedTiles[pp.x][pp.y] = Tile(type: .player(playerData))
         
         tiles = updatedTiles
         
         return Transformation(transformation: transformation.tileTransformation,
-                              inputType: .collectItem(coord, item),
+                              inputType: .collectItem(coord, item, data.carry.total(in: .gold)),
```

Comment:
Here we always insert the .gold, but it should just add the amount of the collected item type